# Documentation

* [Installation](#installation)
* [Prerequisites](#prerequisites)
* [Usage](#usage)
    * [Input](#input)
    * [Options](#options)
    * [Output](#output)
* [Informatic resources](#informatic-resources)
    * [Memory](#memory)
    * [Cpu](#cpu)
* [Examples](#examples)
    * [Standard](#standard)
    * [Slurm](#slurm)
    * [PBS/Torque](#pbstorque)
* [Performance](#performance)
* [Parallel filesystems](#parallel-filesystems)
* [Algorithm](#algorithm)
* [References](#references)


## Installation

Follow the [Installation](INSTALL.md) guidelines to compile and install `mpiBWA`.

## Prerequisites

As `mpiBWA` relies on the Message Passing Interface (MPI) standard, `mpirun` must be available to run the program. Several MPI implementations exist
such as [mpich](https://www.mpich.org/), [open-mpi](https://www.open-mpi.org/) or [IntelÂ® MPI Library](https://software.intel.com/en-us/mpi-library). The `mpirun` program must be available in your PATH.

### Install mpirun on CentOS

* for [open-mpi](https://www.open-mpi.org/): `sudo yum install openmpi openmpi-devel`
* for [mpich](https://www.mpich.org/): `sudo yum install mpich`

### Install mpirun on Ubuntu

* for [open-mpi](https://www.open-mpi.org/): `sudo apt-get install openmpi-bin`
* for [mpich](https://www.mpich.org/): `sudo apt-get install mpich`

## Usage

After installation MAkefile creates 2 binaries: mpiBWA and mpiBWAIdx

`mpiBWAIdx` is responsible for creating a binary image of the reference genome. This image is subsequently loaded in shared memory by `mpiBWA`. Then every `mpiBWA` process on the computing node will share the same genome reference. mpiBWAIdx deos'nt need MPI to run. To create an image type:

`mpiBWAIdx my_ref.fa` this will create the file `my_ref.fa.map`. `mpiBWAIdx` requires the following files `my_ref.fa.sa`, `my_ref.fa.bwt`, `my_ref.fa.ann`, `my_ref.fa.pac`, `my_ref.fa.amb` to construct the `fa.map` file. Those file are generated by `bwa index`. It works also with the fasta.  

`mpiBWA` is executed with the `mpirun` program, for example:

`mpirun -n 1 mpiBWA mem -t 8 -o $SAM_OUT $REFERENCE $FASTQ1 $FASTQ2` will launch 1 process MPI and 8 threads will be created by mpiBWA.

The `-n` options passed to `mpirun` indicates the number of processes to run in parallel (this is basically the number of cores that will be used). For for details on how to choose the number processes, see the [Informatic resources](#informatic-resources) section.

`mpiBWA` requires four mandatory arguments:

* [Input](#input): a reference genome in .map format, one or two fastq files trimmed or not 
* [Output](#output): a sam file 

## Informatic resources

### Memory

The total memory used during the alignment if approximately the size of the .map plus the size of the sam loaded by each bwa tasks. A BWA threads consume around 300 MB.
 





